osquery/bro/BrokerManager.cpp

```diff
@@ -0,0 +1,312 @@
+  this->sendEvent(TOPIC_ANNOUNCE, announceMsg);
+
+  return Status(0, "OK");
+}
+
+/////////////////////////////////////////////////////////
+//////////////// Broker Send Methods/////////////////////
+/////////////////////////////////////////////////////////
+
+Status BrokerManager::logQueryLogItemToBro(const QueryLogItem& qli) {
+  // Attributes from QueryLogItem
+  std::string queryID = qli.name; // The QueryID
+  //  std::string identifier = qli.identifier; // The HostID
+  //  size_t time = qli.time;
+  //  std::string calendar_time = qli.calendar_time;
+  //  std::map<std::string, std::string> decorations = qli.decorations;me;
+
+  // Is this schedule or one-time? Get Query and Type
+  std::string query;
+  std::string qType;
+  auto status_find = this->qm->findQueryAndType(queryID, qType, query);
+  if (!status_find.ok()) {
+    return status_find;
+  }
+
+  // Rows to be reported
+  std::vector<std::tuple<osquery::Row, std::string>> rows;
+  for (const auto& row : qli.results.added) {
+    rows.emplace_back(row, "ADD");
+  }
+  for (const auto& row : qli.results.removed) {
+    rows.emplace_back(row, "REMOVE");
+  }
+  for (const auto& row : qli.snapshot_results) {
+    rows.emplace_back(row, "SNAPSHOT");
+  }
+
+  // Get Info about SQL Query and Types
+  TableColumns columns;
+  Status status = getQueryColumns(query, columns);
+  std::map<std::string, ColumnType> columnTypes;
+  for (std::tuple<std::string, ColumnType, ColumnOptions> t : columns) {
+    std::string columnName = std::get<0>(t);
+    ColumnType columnType = std::get<1>(t);
+    //      ColumnOptions columnOptions = std::get<2>(t);
+    columnTypes[columnName] = columnType;
+    //    LOG(INFO) << "Column named '" << columnName << "' is of type '" <<
+    //    kColumnTypeNames.at(columnType) << "'";
+  }
+
+  // Common message fields
+  std::string uid = this->getNodeID();
+  std::string topic = this->qm->getEventTopic(queryID);
+  std::string event_name = this->qm->getEventName(queryID);
+  LOG(INFO) << "Creating " << rows.size() << " messages for events with name :'"
+            << event_name << "'";
+
+  // Create message for each row
+  bool parse_err = false;
+  for (const auto& element : rows) {
+    // Get row and trigger
+    osquery::Row row = std::get<0>(element);
+    std::string trigger = std::get<1>(element);
+
+    // Set event name, uid and trigger
+    broker::message msg;
+    msg.push_back(event_name);
+    broker::record result_info({broker::record::field(broker::data(uid)),
+                                broker::record::field(broker::data(
+                                    broker::enum_value{"osquery::" + trigger})),
+                                broker::record::field(broker::data(
+                                    this->qm->getEventCookie(queryID)))});
+    msg.push_back(result_info);
+
+    // Format each column
+    for (std::tuple<std::string, ColumnType, ColumnOptions> t : columns) {
```

Comment:
The nested `for`s are relevant for error conditions only. They give some detailed information about the cause. Meanwhile I identified the reason why this error happened to me. I would keep this code in case related errors occur